documentation/secure_storage.md

```diff
@@ -0,0 +1,223 @@
+- **core/tee/tee_fs_common.c:** TEE file system & REE file operation interface
+- **core/tee/tee_fs_key_manager.c:** Key manager
+- **lib/libutee/:** GlobalPlatform Internal API library
+
+### Basic File Operation Flow
+
+When TA is calling the write function provided by GP Trusted Storage API to
+write data to a persistent object, a corresponding syscall implemented in TEE
+Trusted Storage Service will be called, which in turn will invoke a series of
+TEE file operations to store the data. TEE file system will then encrypt the
+data (when CFG_ENC_FS=y) and send REE file operation commands and the encrypted
+data to TEE supplicant by a series of RPC messages. TEE supplicant will receive
+the messages and help to store the encrypted data accordingly to the Linux file
+system. Reading files are handled in a similar manner.
+
+### GlobalPlatform Trusted Storage Requirement
+
+Below is an excerpt from the specification listing the most vital requirements:
+
+1. The Trusted Storage may be backed by non-secure resources as long as suitable
+   cryptographic protection is applied, which MUST be as strong as the means
+   used to protect the TEE code and data itself
+2. The Trusted Storage MUST be bound to a particular device, which means that it
+   MUST be accessible or modifiable only by authorized TAs running in the same
+   TEE and on the same device as when the data was created.
+3. Ability to hide sensitive key material from the TA itself.
+4. Each TA has access to its own storage space that is shared among all the
+   instances of that TA but separated from the other TAs.
+5. The Trusted Storage must provide a minimum level of protection against
+   rollback attacks. It is accepted that the actually physical storage may be in
+   an unsecure areas and so is vulnerable to actions from outside of the TEE.
+   Typically, an implementation may rely on the REE for that purpose (protection
+   level 100) or on hardware assets controlled by the TEE (protection level
+   1000).
+
+### TEE File Structure In Linux File System
+
+![TEE File Structure](images/secure_storage/tee_file_structure.png
+"TEE file structure in Linux file system")
+
+OP-TEE by default use "/data/tee/" as the secure storage space in Linux
+file system. For each TA, OP-TEE use the TA's UUID to create a standalone folder
+for it under the secure storage space folder. For a persistent object belonging
+to a specific TA, OP-TEE creates a TEE file folder which's name is object-id
+under the TA folder.
+
+In TEE file folder, there is a meta file and several block files. Meta file is
+for storing the information of the TEE file which is used by TEE file system to
+manage the TEE file; block file is for storing the data of the persistent
+object.
+
+If the compile time flag CFG_ENC_FS is set to 'y', the  data stored in block
+files will be encrypted, otherwise, the data will not be encrypted. The
+information stored in meta file are always encrypted.
```

Comment:
I imagine people will ask: What is the reason for that? I.e, always having the meta data encrypted?